Sources/AsyncHTTPClient/RedirectState.swift

```diff
@@ -0,0 +1,140 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import struct Foundation.URL
+import NIOHTTP1
+
+struct RedirectState {
+    /// number of redirects we are allowed to follow.
+    private var count: Int
+
+    /// All visited URLs.
+    private var visited: [String]
+
+    /// if true, `redirect(to:)` will throw an error if a cycle is detected.
+    private let allowCycles: Bool
+}
+
+extension RedirectState {
+    /// Creates a `RedirectState` from a configuration.
+    /// Returns nil if the user disallowed redirects,
+    /// otherwise an instance of `RedirectState` which respects the user defined settings.
+    init?(
+        _ configuration: HTTPClient.Configuration.RedirectConfiguration.Configuration,
```

Comment:
This type name is _awful_. Is it possible we could give some of these types clearer names? If not, can we use a `typealias`?

Reply:
sadly `HTTPClient.Configuration.RedirectConfiguration` is public API but we can change the last `Configuration` to e.g. `Mode`. In addition, we can add a typealias:
```swift
internal typealias RedirectMode = HTTPClient.Configuration.RedirectConfiguration.Mode
```
WDYT?

Reply:
I'd like both of those changes, great idea!